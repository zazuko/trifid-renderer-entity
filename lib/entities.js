import rdf from 'rdf-ext'
import { subjects } from './builder/utils.js'

function entities (cf, entityBuilder) {
  const visited = rdf.termSet()
  const entities = []

  // If this clownface is pointing  to a term, render it first
  const allSubjects = cf.term ? [cf.term, ...subjects(cf.any()).filter(node => !node.equals(cf.term))] : subjects(cf.any())

  for (const subject of [...allSubjects]) {
    if (!visited.has(subject)) {
      const { entity } = entityBuilder.withCf(cf.node(subject)).withVisited(visited).buildWithContext()
      entities.push(entity)
    }
  }
  return entities
}

export { entities }
