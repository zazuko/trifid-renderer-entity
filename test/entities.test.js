import chai from 'chai'
import expect from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { describe, it } from 'mocha'
import rdf from 'rdf-ext'
import { entityBuilder } from '../lib/builder/entityBuilder.js'
import { entities } from '../lib/entities.js'
import { ns } from '../lib/namespaces.js'
import { toQuads } from './support/serialization.js'

expect.extend({ toMatchSnapshot })

const assert = chai.assert

function clown (turtle, term) {
  return rdf.clownface({ dataset: rdf.dataset().addAll(toQuads(turtle)), term: term })
}

describe('entities', () => {
  it('should be a function', () => {
    assert.equal(typeof entities, 'function')
  })
})

describe('tests', () => {
  it('do not repeat without compact', function () {
    const data = `
<a> <d> <e> ;
    <d> <e> .
<e> <b> <c>.
`
    const cf = clown(data, rdf.namedNode('a'))

    const compactMode = false

    const builder = entityBuilder(cf)
      .embedLists(compactMode)
      .groupValuesByProperty(compactMode)
      .groupPropertiesByValue(compactMode)
      .withExternalLabels(rdf.clownface({ dataset: rdf.dataset() }))
      .withPreferredLanguages(['de', 'fr', 'it', 'en'])

    const result = entities(cf, builder)

    expect(result).toMatchSnapshot(this)
  })

  it('do not repeat with compact', function () {
    const data = `
<a> <d> <e> ;
    <d> <e> .
<e> <b> <c>.
`

    const cf = clown(data, rdf.namedNode('a'))

    const compactMode = false

    const builder = entityBuilder(cf)
      .embedLists(compactMode)
      .groupValuesByProperty(compactMode)
      .groupPropertiesByValue(compactMode)
      .withExternalLabels(rdf.clownface({ dataset: rdf.dataset() }))
      .withPreferredLanguages(['de', 'fr', 'it', 'en'])

    const result = entities(cf, builder)

    expect(result).toMatchSnapshot(this)
  })

  it('fetch languages', function () {
    const data = `

<a> <${ns.schema.name}> "Or me" ;
    <${ns.schema.name}> "Me"@de .
`
    const cf = clown(data, rdf.namedNode('a'))

    const builder = entityBuilder(cf)
      .withExternalLabels(rdf.clownface({ dataset: rdf.dataset() }))
      .withPreferredLanguages(['de', 'fr', 'it', 'en'])

    const result = entities(cf, builder)

    expect(result).toMatchSnapshot(this)
  })
})
