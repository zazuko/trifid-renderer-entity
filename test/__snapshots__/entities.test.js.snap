// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests do not repeat with compact 1`] = `
Array [
  Object {
    "label": Object {
      "string": "a",
    },
    "rows": Array [
      Object {
        "properties": Array [
          Object {
            "label": Object {
              "string": "d",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "d",
            },
          },
        ],
        "values": Array [
          Object {
            "label": Object {
              "string": "e",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "e",
            },
          },
        ],
      },
    ],
    "term": NamedNode {
      "value": "a",
    },
  },
  Object {
    "label": Object {
      "string": "e",
    },
    "rows": Array [
      Object {
        "properties": Array [
          Object {
            "label": Object {
              "string": "b",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "b",
            },
          },
        ],
        "values": Array [
          Object {
            "label": Object {
              "string": "c",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "c",
            },
          },
        ],
      },
    ],
    "term": Object {
      "termType": "NamedNode",
      "value": "e",
    },
  },
]
`;

exports[`tests do not repeat without compact 1`] = `
Array [
  Object {
    "label": Object {
      "string": "a",
    },
    "rows": Array [
      Object {
        "properties": Array [
          Object {
            "label": Object {
              "string": "d",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "d",
            },
          },
        ],
        "values": Array [
          Object {
            "label": Object {
              "string": "e",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "e",
            },
          },
        ],
      },
    ],
    "term": NamedNode {
      "value": "a",
    },
  },
  Object {
    "label": Object {
      "string": "e",
    },
    "rows": Array [
      Object {
        "properties": Array [
          Object {
            "label": Object {
              "string": "b",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "b",
            },
          },
        ],
        "values": Array [
          Object {
            "label": Object {
              "string": "c",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "c",
            },
          },
        ],
      },
    ],
    "term": Object {
      "termType": "NamedNode",
      "value": "e",
    },
  },
]
`;

exports[`tests fetch languages 1`] = `
Array [
  Object {
    "label": Object {
      "language": "de",
      "string": "Me",
    },
    "rows": Array [
      Object {
        "properties": Array [
          Object {
            "label": Object {
              "string": "name",
              "vocab": "schema",
            },
            "term": Object {
              "termType": "NamedNode",
              "value": "http://schema.org/name",
            },
          },
        ],
        "values": Array [
          Object {
            "label": Object {
              "datatype": Object {
                "string": "string",
                "vocab": "xsd",
              },
              "string": "Or me",
            },
            "term": Object {
              "datatype": Object {
                "termType": "NamedNode",
                "value": "http://www.w3.org/2001/XMLSchema#string",
              },
              "language": "",
              "termType": "Literal",
              "value": "Or me",
            },
          },
          Object {
            "label": Object {
              "datatype": Object {
                "string": "langString",
                "vocab": "rdf",
              },
              "language": "de",
              "string": "Me",
            },
            "term": Object {
              "datatype": Object {
                "termType": "NamedNode",
                "value": "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString",
              },
              "language": "de",
              "termType": "Literal",
              "value": "Me",
            },
          },
        ],
      },
    ],
    "term": NamedNode {
      "value": "a",
    },
  },
]
`;
